Fix to send ROLLBACK PREPARED when parallel_commit is on

YKatsuragi opened the pull request:
On a transaction abort, the current implementation might not send ROLLBACK PREPARED to those connections that have sent PREPARED TRANSACTION with parallel_commit=on.
Because PQgetResult for PREPARED TRANSACTION is not called and xact_changing_state is true for those connecctions, pgfdw_abort_cleanup_with_sql skips to send ROLLBACK PREPARED.

This PR fixes this by cleaning up the pending_entries_prepare at the beginning of the abort process of pgfdw_xact_two_phase. Along with this fix on parallel commit, this PR fixes the regression test to keep the number of prepared transactions below max_prepared_xacts.

MasaoFujii commented:
@YKatsuragi 

> On a transaction abort, the current implementation might not send ROLLBACK PREPARED to those connections that have sent PREPARED TRANSACTION with parallel_commit=on.
Because PQgetResult for PREPARED TRANSACTION is not called and xact_changing_state is true for those connecctions, pgfdw_abort_cleanup_with_sql skips to send ROLLBACK PREPARED.

Thanks for reporting the issue! Your analysis of the problem seems accurate.

> This PR fixes this by cleaning up the pending_entries_prepare at the beginning of the abort process of pgfdw_xact_two_phase. 

Thank you for providing the patch! I have some review comments for it.

```
+/*
+ * This list has ConnCacheEntry that are parallel_commit=on and have already
+ * sent PREPARED TRANSACTION. On a transaction abort, PQgetResult should be
+ * called to these connections before sending ROLLBACK PREPARED.
+ */
+
List	   *pending_entries_prepare = NIL;
```

Shouldn't we declare this pointer as static since it's only used in postgres_fdw_plus.c?

```
+	if ((event == XACT_EVENT_ABORT || event == XACT_EVENT_PARALLEL_ABORT)
+		&& pending_entries_prepare)
+		pgfdw_cleanup_pending_entries();
```

Wouldn't it be simpler to merge this code into the "switch" code block at the top of pgfdw_xact_two_phase()?

If event is XACT_EVENT_ABORT and pending_entries_prepare is NIL, shouldn't pgfdw_xact_two_phase() simply return without performing any actions, and let pgfdw_xact_callback() handle this case instead?

Could you add a brief code comment explaining the necessity of pgfdw_cleanup_pending_entries() when pending_entries_prepare is not NIL and event is XACT_EVENT_ABORT? This would make it easier for other developers to understand the code.

```
+	if (pending_entries_prepare)
+	{
+		list_free(pending_entries_prepare);
+		pending_entries_prepare = NIL;
+	}
```

Is list_free() necessary in this case, or is it not required since the memory context mechanism takes care of freeing the pending_entries_prepare?

```
-					pgfdw_prepare_xacts(entry, &pending_entries_prepare);
+					pgfdw_prepare_xacts(entry);
```

I don't think it's necessary to change the declarations of function arguments. How about leaving the functions that the patch removed the argument pending_entries_prepare as they are and making the global variable "pending_entries_prepare" a static variable inside pgfdw_xact_two_phase()? Wouldn't this approach be simpler?

```
+		endtime = TimestampTzPlusMilliseconds(GetCurrentTimestamp(),
+											  CONNECTION_CLEANUP_TIMEOUT);
+		PreparedXactCommand(sql, "ROLLBACK PREPARED", entry);
```

"ROLLBACK PREPARED" here should be "PREPARE TRANSACTION"?

YKatsuragi commented:
Thank you for reviewing!

> ```
> +	if ((event == XACT_EVENT_ABORT || event == XACT_EVENT_PARALLEL_ABORT)
> +		&& pending_entries_prepare)
> +		pgfdw_cleanup_pending_entries();
> ```
> Wouldn't it be simpler to merge this code into the "switch" code block at the top of pgfdw_xact_two_phase()?

Moved this code to switch code block.

> If event is XACT_EVENT_ABORT and pending_entries_prepare is NIL, shouldn't pgfdw_xact_two_phase() simply return without performing any actions, and let pgfdw_xact_callback() handle this case instead?

I think no. In that case, postgres_fdw_plus should keep processing and send ROLLBACK PREPARED to connections (these connections are not necessarily parallel_commit=on).

> Could you add a brief code comment explaining the necessity of pgfdw_cleanup_pending_entries() when pending_entries_prepare is not NIL and event is XACT_EVENT_ABORT? This would make it easier for other developers to understand the code.

Added a comment.

> ```
> +	if (pending_entries_prepare)
> +	{
> +		list_free(pending_entries_prepare);
> +		pending_entries_prepare = NIL;
> +	}
> ```
> Is list_free() necessary in this case, or is it not required since the memory context mechanism takes care of freeing the pending_entries_prepare?

I think this is not necessary. Removed.

> I don't think it's necessary to change the declarations of function arguments. How about leaving the functions that the patch removed the argument pending_entries_prepare as they are and making the global variable "pending_entries_prepare" a static variable inside pgfdw_xact_two_phase()? Wouldn't this approach be simpler?

Changed pending_entries_prepare to function local static variable and keep the functions' argument.

> ```
> +		endtime = TimestampTzPlusMilliseconds(GetCurrentTimestamp(),
> +											  CONNECTION_CLEANUP_TIMEOUT);
> +		PreparedXactCommand(sql, "ROLLBACK PREPARED", entry);
> ```
> "ROLLBACK PREPARED" here should be "PREPARE TRANSACTION"?

Fixed.

YKatsuragi commented:
> > > If event is XACT_EVENT_ABORT and pending_entries_prepare is NIL, shouldn't pgfdw_xact_two_phase() simply return without performing any actions, and let pgfdw_xact_callback() handle this case instead?
> > 
> > I think no. In that case, postgres_fdw_plus should keep processing and send ROLLBACK PREPARED to connections (these connections are not necessarily parallel_commit=on).
> 
> Yes, you're right! BTW, what about the case where the event is XACT_EVENT_ABORT/XACT_EVENT_PARALLEL_ABORT and pgfdw_two_phase_commit=false?

postgres_fdw_plus calls pgfdw_abort_cleanup to all connections to send ABORT command in pgfdw_xact_two_phase.
This PR does not change the behavior of the case and this behavior is the same as postgres_fdw (now postgres_fdw has parallel_abort though).
It seems better for postgres_fdw_plus to do nothing and go back to postgres_fdw if pgfdw_two_phase_commit.
I would like to modify like the following, what do you think?

```
        case XACT_EVENT_ABORT:
        case XACT_EVENT_PARALLEL_ABORT:
            if (!pgfdw_two_phase_commit)
                return false;
            if (pending_entries_prepare)
                pgfdw_cleanup_pending_entries(pending_entries_prepare);
            break;
```

MasaoFujii commented:
@YKatsuragi 
Thanks for updating the patch!

The recent commit 7ec45c7 updated the main branch. Could you please rebase the patch?

> +/*
> + * This list has ConnCacheEntry that are parallel_commit=on and have already
> + * sent PREPARED TRANSACTION. On a transaction abort, PQgetResult should be
> + * called to these connections before sending ROLLBACK PREPARED.
> + */
> +List	   *pending_entries_prepare = NIL;
> +

Since pending_entries_prepare is already declared as a static variable inside pgfdw_xact_two_phase(), isn't it necessary to have an additional declaration?

> +extern void pgfdw_cleanup_pending_entries(List *pending_entries_prepare);

Since pgfdw_cleanup_pending_entries() is only called in postgres_fdw_plus.c, do we really need to mark this function as an extern?

> +	if (pending_entries_prepare)
> +		pending_entries_prepare = NIL;

I suggest setting pending_entries_prepare to NIL only after the list entries are actually cleaned up, such as after pgfdw_cleanup_pending_entries() or pgfdw_finish_prepare_cleanup() is called. Additionally, it might be beneficial to add an assertion check for impossible combinations, such as throwing an assertion failure if the event is XACT_EVENT_COMMIT and pending_entries_prepare != NIL. What do you think about this approach?

> +		/*
> +		 * Entries in the pending_entries_prepare are parallel_commit=on and
> +		 * have already sent PREPARE TRANSACTION, however, PQgetResult might
> +		 * still not be called to some of them. It is necessary to call
> +		 * PQgetResult before sending ROLLBACK PREPARED.
> +		 */

This comment seems difficult to be understood. How about something like the following?

```
When the event is XACT_EVENT_ABORT or XACT_EVENT_PARALLEL_ABORT, it's possible that there are pending connection entries where PREPARE TRANSACTION commands were sent to remote servers but their responses haven't been received yet. This situation can occur if, for example, an error is encountered during the execution of PREPARE TRANSACTION commands in pgfdw_xact_two_phase() with parallel_commit = on. To ensure proper handling, we make an effort to receive these responses before issuing ROLLBACK PREPARED commands. This is necessary because once we receive all the results from the last command, like ROLLBACK PREPARED, we cannot issue any new commands.
```

YKatsuragi commented:
Thank you for reviewing!

> The recent commit https://github.com/pgfdwplus/postgres_fdw_plus/commit/7ec45c7ddbf433c56d676db2947498d812bfae7a updated the main branch. Could you please rebase the patch?

Rebased.

> Since pending_entries_prepare is already declared as a static variable inside pgfdw_xact_two_phase(), isn't it necessary to have an additional declaration?

Yes, I forgot to remove. Removed.

> Since pgfdw_cleanup_pending_entries() is only called in postgres_fdw_plus.c, do we really need to mark this function as an extern?

I followed the existing coding style and marked them as extern (like `pgfdw_xact_two_phase` and `pgfdw_arrange_read_committed`).
Do you mean because these functions are only called in postgres_fdw_plus.c, move their declarations to postgres_fdw_plus.c and mark them as static?

> I suggest setting pending_entries_prepare to NIL only after the list entries are actually cleaned up, such as after pgfdw_cleanup_pending_entries() or pgfdw_finish_prepare_cleanup() is called. Additionally, it might be beneficial to add an assertion check for impossible combinations, such as throwing an assertion failure if the event is XACT_EVENT_COMMIT and pending_entries_prepare != NIL. What do you think about this approach?

Seems better to set NIL only after cleaning up the `pending_entries_prepare`.
Fixed so. Also, added an assertion.
Since `pgfdw_cleanup_pending_entries()` set NIL to `pending_entries_prepare` in abort case, `pending_entries_prepare` should always be NIL after passing the switch statement. The assertion checks that.

> This comment seems difficult to be understood. How about something like the following?

Fixed as you proposed, but modified a little bit.

MasaoFujii commented:
@YKatsuragi 
Thanks for updating the patch!

> I followed the existing coding style and marked them as extern (like pgfdw_xact_two_phase and pgfdw_arrange_read_committed).

The reason they were marked as extern is that they are defined in postgres_fdw_plus.c but called in connection.c.

> Do you mean because these functions are only called in postgres_fdw_plus.c, move their declarations to postgres_fdw_plus.c and mark them as static?

I suggest defining pgfdw_cleanup_pending_entries() as a static function in postgres_fdw_plus.c since it is only used there.

If there are other functions marked as extern that are also used only in postgres_fdw_plus.c, they should also be defined as static functions. If any such functions exist, it would be better to make this change in a separate patch.

> Seems better to set NIL only after cleaning up the pending_entries_prepare.
> Fixed so. Also, added an assertion.
> Since pgfdw_cleanup_pending_entries() set NIL to pending_entries_prepare in abort case, pending_entries_prepare should always be NIL after passing the switch statement. The assertion checks that.

Many thanks!

Here are other review comments:

> ```
>   					continue;
>  -
>   				case XACT_EVENT_PARALLEL_COMMIT:
> ```

Why was this empty line removed?

> ```
>  	ListCell   *lc;
>  
> -	Assert(pending_entries_prepare);
> +	Assert(*pending_entries_prepare);
> ```

Should we consider adding the same assertion check to pgfdw_cleanup_pending_entries() as well? It would be beneficial to have similar code in both pgfdw_finish_prepare_cleanup() and pgfdw_cleanup_pending_entries() functions.

MasaoFujii commented:
@YKatsuragi 
The main problem with the pgfdw_plus_loopback2 regression test and parallel_commit = on is the instability of the following query results, correct? Depending on whether the commit phase is processed in the order of loopback1 and then loopback2, or vice versa, the number of prepared foreign transactions generated varies. This instability is due to the order in which foreign servers are processed. To address this, you temporarily enabled parallel_commit even in loopback1. Do I understand this correctly?

> ```
> BEGIN;
> INSERT INTO ft1 VALUES (210);
> INSERT INTO ft2 VALUES (210);
> SELECT pg_terminate_backend(pid, 10000) FROM pg_stat_activity
>     WHERE application_name = 'pgfdw_plus_loopback1';
> COMMIT;
> ```

However, as an alternative solution to the problem, how about simply removing the unstable test case mentioned above? If we do that, it seems that we can avoid the additional following changes you made to the regression test. What do you think?

> ```
> --- Create one foreign prepared transaction that should be rollbacked
> +-- Resolve another prepared transaction to keep the number of prepared transactions
> +-- below max_prepared_xacts
> +SELECT status, count(*)
> +  FROM pgfdw_plus_resolve_foreign_prepared_xacts('pgfdw_plus_loopback2')
> +  GROUP BY status ORDER BY status;
> +SELECT count(*) FROM pgfdw_plus_foreign_prepared_xacts('pgfdw_plus_loopback1');
> +SELECT count(*) FROM pgfdw_plus_foreign_prepared_xacts('pgfdw_plus_loopback2');
> +
> +-- Create two foreign prepared transactions that should be rollbacked
>  -- so as to test later whether pgfdw_plus_resolve_foreign_prepared_xacts_all()
> --- can actually rollback it.
> +-- can actually rollback them.
> +-- Temporarily set parallel_commit to 'on' in pgfdw_plus_loopaback1 to surely
> +-- create prepared transactions.
> +ALTER SERVER pgfdw_plus_loopback1 OPTIONS (add parallel_commit 'on');
> ```

YKatsuragi commented:
> > I followed the existing coding style and marked them as extern (like pgfdw_xact_two_phase and pgfdw_arrange_read_committed).
> 
> The reason they were marked as extern is that they are defined in postgres_fdw_plus.c but called in connection.c.
> ...
> I suggest defining pgfdw_cleanup_pending_entries() as a static function in postgres_fdw_plus.c since it is only used there.

Sorry, these were not good as examples. (`pgfdw_prepare_xacts` is an example that is used only in postgres_fdw_plus.c but not marked as static.)
I fixed to make `pgfdw_cleanup_pending_entries` as a static function.


> If there are other functions marked as extern that are also used only in postgres_fdw_plus.c, they should also be defined as static functions. If any such functions exist, it would be better to make this change in a separate patch.

Yes. I agree with you.

> Why was this empty line removed?

Fixed.


> Should we consider adding the same assertion check to pgfdw_cleanup_pending_entries() as well? It would be beneficial to have similar code in both pgfdw_finish_prepare_cleanup() and pgfdw_cleanup_pending_entries() functions.

Added.


> The main problem with the pgfdw_plus_loopback2 regression test and parallel_commit = on is the instability of the following query results, correct? Depending on whether the commit phase is processed in the order of loopback1 and then loopback2, or vice versa, the number of prepared foreign transactions generated varies. This instability is due to the order in which foreign servers are processed. To address this, you temporarily enabled parallel_commit even in loopback1. Do I understand this correctly?

Yes.

> However, as an alternative solution to the problem, how about simply removing the unstable test case mentioned above? If we do that, it seems that we can avoid the additional following changes you made to the regression test. What do you think?

This way seems simpler. So I fixed to remove unstable test case and additional changes.
